In [1]:
import os
import sys
import numpy as np
import pandas as pd
import json
import random

In [2]:
train_file = '/data/sulixin/research/contest/WikiQACodePackage/data/wiki/WikiQASent-train.txt'
dev_file =  '/data/sulixin/research/contest/WikiQACodePackage/data/wiki/WikiQASent-dev.txt'
test_file = '/data/sulixin/research/contest/WikiQACodePackage/data/wiki/WikiQASent-test.txt'

In [8]:
def load_file(data_file):
    rv = []
    
    for line in open(data_file):
        data = line.strip().split('\t')
        assert len(data) == 3
        assert data[-1] in set(['0', '1'])
        rv.append({'query': data[0], 'passage': data[1], 'label': data[-1]})
    return rv 
train = load_file(train_file)
dev = load_file(dev_file)
test = load_file(test_file)

In [39]:
def analysis_by_pandas(data, label=None):
    print('\n========{}======='.format(label))
    df = pd.DataFrame(data)
    df['passage_len'] = df['passage'].apply(lambda x:len(x.split()))
    df['query_len'] = df['query'].apply(lambda x:len(x.split()))
    df['label_int'] = df['label'].apply(lambda x:int(x))
    print(df.describe())
    gdf = df.groupby(['query']).sum()['label_int'].value_counts(normalize=False,ascending=True)
    #print(gdf)
    gdf = df.groupby(['query']).mean()
    print(gdf.describe())
    gdf = df.groupby(['query']).mean()
    #print(gdf)
    print(gdf[gdf['label_int']!=0].describe())
    gdf = df.groupby(['query']).count()
    print(gdf.describe())
analysis_by_pandas(train, 'train')
analysis_by_pandas(dev, 'dev')
analysis_by_pandas(test, 'test')


========train=======
        passage_len     query_len     label_int
count  20360.000000  20360.000000  20360.000000
mean      25.290766      7.117534      0.051081
std       13.019002      2.578649      0.220167
min        1.000000      2.000000      0.000000
25%       16.000000      5.000000      0.000000
50%       23.000000      7.000000      0.000000
75%       32.000000      8.000000      0.000000
max      236.000000     23.000000      1.000000
       passage_len    query_len    label_int
count  2118.000000  2118.000000  2118.000000
mean     25.300579     7.156280     0.079204
std       6.786060     2.570615     0.147173
min       2.250000     2.000000     0.000000
25%      21.166667     5.000000     0.000000
50%      24.785714     7.000000     0.000000
75%      28.692308     8.000000     0.111111
max     111.000000    23.000000     1.000000
       passage_len   query_len   label_int
count   873.000000  873.000000  873.000000
mean     25.779438    6.454754    0.192158
std       7.

数据汇总

| train      |     dev |   test   |
| :-------- | --------:| :------: |
|1245/2118 (no answer/all query)   |   170/296 |  390/633  |
| 20360(passages)    |   2733 |  6165  |
| 9.61(passages/query)    |   9.233 |   9.73  |
| 25.3(words/passage)    |  24.6 |  24.95  |
| 7.1(words/query)    |   7.23 |  7.25  |
| 0.192(positive/passages on answerable)    |   0.21 |  0.20  |